File: src/com/medium/ValidateBinarySearchTree_98.java
package com.medium;

import java.util.ArrayList;
import java.util.List;

/**
 * @author Chris Lin
 * @version 1.0
 */
public class ValidateBinarySearchTree_98 {
    // 2019年9月12日
    // Use Recursive
    // Time Complexity: O(n)
    // Space Complexity:O(n)
    // Runtime: 1 ms, faster than 48.17%
    // Memory Usage: 39.3 MB, less than 80.47%
    public boolean isValidBST(TreeNode root) {
        return checkBST(root, null, null);
    }

    private boolean checkBST(TreeNode root, Integer mMin, Integer mMax) {
        if (root == null) {
            return true;
        }
        int mValue = root.val;
        if (mMin != null && mValue <= mMin) {
            return false;
        }
        if (mMax != null && mValue >= mMax) {
            return false;
        }
        if (!checkBST(root.left, mMin, mValue)) {
            return false;
        }
        if (!checkBST(root.right, mValue, mMax)) {
            return false;
        }

        return true;
    }

    // 2019年9月12日
    // Use inorder traversal
    // Time Complexity: O(n)
    // Space Complexity:O(n)
    // Runtime: 2 ms, faster than 24.31%
    // Memory Usage: 37.2 MB, less than 100.00%
    private List<Integer> mList;

    public boolean isValidBST2(TreeNode root) {
        mList = new ArrayList<>();
        copyBST(root);
        for (int i = 0; i < mList.size() - 1; i++) {
            if (mList.get(i) >= mList.get(i + 1)) {
                return false;
            }
        }
        return true;
    }

    private void copyBST(TreeNode root) {
        if (root == null) {
            return;
        }
        copyBST(root.left);
        mList.add(root.val);
        copyBST(root.right);
    }
}

File: Golang_src/medium/validateBinarySearchTree_98.go
package medium

import "math"

// Use recursive:
// Time Complexity: O(n)
// Space Complexity:O(n)
// Runtime: 32 ms, faster than 8.90%
// Memory Usage: 5.6 MB, less than 28.23%
func isValidBST(root *TreeNode) bool {
	return helpers(root, math.MinInt64, math.MaxInt64)
}

func helpers(root *TreeNode, lower, upper int) bool {
	if root == nil {
		return true
	}
	if root.Val <= lower || root.Val >= upper {
		return false
	}

	return helpers(root.Right, root.Val, upper) && helpers(root.Left, lower, root.Val)
}

// Use inorder traversal:
// Time Complexity: O(n)
// Space Complexity:O(n)
// Runtime: 8 ms, faster than 44.37%
// Memory Usage: 5.8 MB, less than 12.34%
func isValidBST2(root *TreeNode) bool {
	stack := []*TreeNode{}
	cur := math.MinInt64
	for len(stack) > 0 || root != nil {
		for root != nil {
			stack = append(stack, root)
			root = root.Left
		}
		root = stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		if root.Val <= cur {
			return false
		}
		cur = root.Val
		root = root.Right
	}
	return true
}
